packages/material-ui/src/OverridableComponent.d.ts

```diff
@@ -0,0 +1,39 @@
+import * as React from 'react';
+import { ConsistentWith, PropsOf, Omit } from '.';
+import { StyledComponentProps } from './styles';
+
+export interface OverridableComponent<M extends OverridableTypeMap> {
+  <C extends AnyReactType>(props: { component: C } & OverrideProps<M, C>): JSX.Element;
+  (props: DefaultProps<M>): JSX.Element;
+}
+
+// prettier-ignore
+export type OverrideProps<
+  M extends OverridableTypeMap,
+  C extends AnyReactType
+> = (
+  & BaseProps<M>
+  & Omit<PropsOf<C>, keyof StyledComponentProps<M['classKey']>>
+);
+
+// prettier-ignore
+export type DefaultProps<M extends OverridableTypeMap> =
```

Comment:
I think this could be named better. I thought this were props with default values but that's not actually the case. This is is the union of the props for the component and the props that can be spread to the "root component", right? ValidProps, FullProps?

If I understood this correctly we should probably add a documentation for that.

Reply:
>This is is the union of the props for the component and the props that can be spread to the "root component", right? ValidProps, FullProps?

It's specifically the union of the base props and the props that the _default_ root component accepts, i.e. the full set of props when no override `component` prop is provided.

I agree it's confusing because of the preexisting meaning of "default props" in React.

`NonOverriddenProps`? 🤢